realm/realm-library/src/main/java/io/realm/RealmInteger.java

```diff
@@ -50,111 +56,67 @@ public RealmInteger(String value) {
 
     /**
      * Decrements the {@code RealmInteger}, subtracting the value of the argument.
      * Increment/decrement from all devices are reflected in the new value, which is guaranteed to converge.
      *
      * @param dec quantity to be subtracted from the counter.
      */
-    public void decrement(long dec) {
-        value -= dec;
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public byte byteValue() {
-        return (byte) value;
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public double doubleValue() {
-        return (double) value;
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public float floatValue() {
-        return (float) value;
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public int intValue() {
-        return (int) value;
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public long longValue() {
-        return value;
-    }
+    public abstract void decrement(long dec);
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public short shortValue() {
-        return (short) value;
+    public String toString() {
+        return String.valueOf(longValue());
     }
 
     /**
-     * {@inheritDoc}
+     * RealmIntegers compare strictly by their values.
+     *
+     * @param o the compare target
+     * @return -1, 0, or 1, depending on whether this object's value is &gt;, =, or &lt; the target's.
      */
     @Override
-    public int compareTo(RealmInteger o) {
-        long otherValue = o.value;
-        return (value == otherValue) ? 0 : ((value > otherValue) ? 1 : -1);
-    }
+    public final int compareTo(RealmInteger o) {
+        long otherValue = o.longValue();
+        long thisValue = longValue();
 
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public String toString() {
-        return String.valueOf(value);
+        return (thisValue == otherValue) ? 0 : ((thisValue > otherValue) ? 1 : -1);
     }
 
     /**
-     * {@inheritDoc}
+     * A RealmInteger's hash code depends only on its value.
+     *
+     * @return true if the target has the same value.
      */
     @Override
-    public int hashCode() {
-        return (int) (value ^ (value >>> 32));
+    public final int hashCode() {
+        long thisValue = longValue();
+        return (int) (thisValue ^ (thisValue >>> 32));
     }
 
     /**
-     * {@inheritDoc}
+     * Two RealmIntegers are {@code .equals} if and only if their longValues are equal.
+     *
+     * @param o compare target
+     * @return true if the target has the same value.
      */
     @Override
-    public boolean equals(Object o) {
+    public final boolean equals(Object o) {
         if (o == this) { return true; }
         if (!(o instanceof RealmInteger)) { return false; }
-        RealmInteger other = (RealmInteger) o;
-        return other.value == value;
+        return longValue() == ((RealmInteger) o).longValue();
```

Comment:
I don't have a lot of experience with these kinds of counters.  I might have the semantics of `.equals` wrong.  My take was that, between `==` and `.equals`, with the semantics I've currently chosen, you get as much information as possible.